### ISSUE TITLE
RFC: DIMER production parity — phased specification and implementation plan

### ISSUE THREAD

**Issue by kurtvalcorza** [kurtvalcorza/tabdpt-regressor-pipeline#12]
## Summary

This RFC proposes a phased plan to bring `tabdpt-regressor-pipeline` to DIMER production parity: an accepted definition of release acceptance, executable verification evidence, operations documentation, and a decision on the validator/finetuner tier split.

It supersedes the informal gap analysis circulated earlier. That analysis was run against local working copies and its baseline column contained material errors, corrected in §2 below. Requirement keywords **MUST**, **MUST NOT**, **SHOULD**, **SHOULD NOT**, and **MAY** are used per RFC 2119/8174.

## 1. Verified current state

Audited at `main` (`60b57af`) with the four baseline pipeline repositories and four satellite repositories cloned and read directly.

| Repository | Tracked files | `src/` | `tests/` | CI |
|---|---:|---|---|---|
| mitra-classifier-pipeline | 35 | no | no | 1 version, `pip install pandas numpy` |
| mitra-regressor-pipeline | 34 | no | no | 1 version |
| **tabdpt-regressor-pipeline** | **19** | **yes** | **yes** | **5 versions (3.10–3.14), installed package, pytest** |
| tabicl-classifier-pipeline | 16 | no | no | 1 version |
| tabicl-regressor-pipeline | 16 | no | no | 1 version |

`tabdpt-regressor-pipeline` is behind on operations documentation and deployable packaging. It is ahead of both baselines on installable packaging, test structure, and CI breadth. Any parity work **MUST NOT** regress those three.

## 2. Corrections to the earlier gap analysis

Recorded because the earlier document circulated with these claims and they are load-bearing for planning.

| Earlier claim | Verified |
|---|---|
| Sample datasets: Mitra "3 zipped", TabICL "1–2 zipped" | Mitra has exactly **one** archive per repo. **TabICL has none** — no `examples/` directory exists. |
| `palmer-penguins.zip`, `telco-customer-churn.zip`, `insurance-medical-charges.zip` | **Zero occurrences** across all four baseline repositories. |
| `examples/build_sample_datasets.py` present in both | **Zero occurrences.** Mitra's script is `examples/build_freshretailnet_dataset.py`. |
| `weights/README.md` present in both | **No `weights/` directory exists in any baseline repository.** Both TabICL READMEs link to `weights/README.md` at line 46; those links are broken. |
| Mitra stores serving context as parquet | Mitra emits **no serving context**. It is a gradient fine-tuner. Only **TabICL** uses `training_context.parquet`. |
| Baseline `DEPLOYMENT.md` documents FlashAttention `sm_80+` / T4 bounds | Not present in either baseline. TabICL's `## GPU` section states only that CUDA is required. |
| TabICL `RELEASE_CHECKLIST` "integrated" | **Correct.** It is `## Pre-enable checklist (release gate)` in `DEPLOYMENT.md`, 9 steps, with step 7 as on-platform serving acceptance. |

### 2.1 The two baselines use opposite layouts

Mitra keeps `validator/` and `finetuner/` as subdirectories **inside** the pipeline repository, and also publishes them as satellite repositories. TabICL's pipeline repository contains neither; its satellites hold the only copy.

The Mitra copies have already drifted. Comparing `mitra-classifier-pipeline/validator/` against `mitra-classifier-dataset-validator/`:

```
validator/validate.py             DIFFERS (20 lines)
validator/Dockerfile              DIFFERS (8 lines)
finetuner/contract_entrypoint.py  DIFFERS (4 lines)
finetuner/Dockerfile              DIFFERS (1 line)
finetuner/Dockerfile.gpu          DIFFERS (3 lines)
validator/validator.py            identical
finetuner/train.py                identical
```

The `validate.py` divergence is presentational — a minified one-liner against a documented expansion — not behavioural. The point stands regardless: nothing keeps the copies synchronised, and this repository has just spent issues #6 and #8 eliminating exactly this failure mode, where one fact had two sources and they disagreed.

## 3. Goals and non-goals

**Goals.** Define release acceptance; add executable verification for the tutorial and artifact-reload paths; document deployment and weights handling; decide and execute the tier split.

**Non-goals.** This RFC does **not** propose changing the pinned TabDPT v1.2 model identity, the `dimer-pipeline.json` parameter schema, in-context fitting semantics, or the existing dataset safety guards.

## 4. Decisions required

Each decision below blocks the phase named. Recommendations are given; the decision is the maintainer's.

**D1 — Tier layout.** Satellite-only (TabICL) or vendored subdirectories plus satellites (Mitra)?
*Recommendation: satellite-only.* The Mitra layout has demonstrably drifted (§2.1), and duplication is the failure mode this repository has been closing.

**D2 — Serving-context format.** Keep `training_context.csv`, or move to `training_context.parquet`?
*Recommendation: adopt parquet in a separate, explicitly versioned change, not folded into another phase.* The justification is consistency with TabICL, the only comparable in-context sibling, whose entire artifact contract is built on the parquet filename (`DIMER_CONTRACT.md:26`, release-gate step 4, `validate_colab_tutorial.py:87`). The fidelity argument offered earlier is largely redundant: the v2 `preprocessing` state already reconstructs dtypes without re-inferring them from the context file. Cost is a format version bump touching `DIMER_CONTRACT.md`, `README.md`, `dimer_runtime.py:410`, and the by-path SHA-256 entries in `result.json` and `artifact.json`, plus a `pyarrow` runtime dependency on a package whose current dependency set is four lightweight libraries.

**D3 — Guard code sharing.** How does a validator repository obtain the dataset guards?
*Recommendation: the validator installs this package from a pinned git tag.* `tabdpt-regressor-pipeline` already implements the full guard surface — `_dataset_limits`, `_normalize_member`, `_validate_archive`, `_validate_direct_dataset`, `_find_csv`, `_zip_member`, `load_dimer_tables`, `_validate_target` — as tested, installable code. Neither baseline can do this because neither is a package. This is the one place where `tabdpt-regressor-pipeline` being ahead should be exploited rather than discarded.

**D4 — Sample dataset.** Which dataset, and under what licence?
*No recommendation.* There is no 3-dataset precedent to match. Mitra ships one domain-specific archive built from FreshRetailNet. The choice **MUST** be made on TabDPT's own needs. `[TODO: maintainer to select dataset and confirm redistribution licence]`

### Phase 0 — Gate: land in-flight work

PR #10 must merge before later phases begin, because it establishes the FlashAttention capability policy and the pre-model target-validation ordering that later phases reference.

- **R0.1** Phase 1 **MUST NOT** begin until PR #10 is merged to `main`.
- **R0.2** Phase 3 documentation **MUST** reference the `## GPU attention compatibility` section of `DIMER_CONTRACT.md` and **MUST NOT** restate the compute-capability threshold.

### Phase 1 — Acceptance definition and tutorial verification

Smallest phase, highest leverage: it defines "done" for every later phase and closes two known regression risks at once.

- **R1.1** The repository **MUST** contain a release gate enumerating the conditions for production enablement, adapted from TabICL's 9-step `## Pre-enable checklist`. It **MAY** live in `DEPLOYMENT.md` rather than a separate file; TabICL demonstrates the integrated form is accepted house practice.
- **R1.2** The release gate **MUST** distinguish repository-verifiable steps from platform-owned steps, and **MUST** state that production enablement is withheld until on-platform serving acceptance passes.
- **R1.3** `scripts/validate_colab_tutorial.py` **MUST** exist and **MUST** parse every code cell of every notebook under `tutorials/` via Python AST, failing on any syntax error.
- **R1.4** The validator **MUST** assert that each notebook instantiating `TabDPTRegressionPipeline` passes `use_flash=False` explicitly.
  *Rationale: `scripts/validate_repo.py` currently checks only `nbformat == 4` and cell count, so the Tesla T4 fix from issue #9 can be reverted with CI green.*
- **R1.5** The validator **MUST** fail on absolute filesystem paths in notebook code cells.
- **R1.6** CI **MUST** invoke `scripts/validate_colab_tutorial.py` on every push and pull request.

**Acceptance.** Deleting `use_flash=False` from a tutorial cell causes CI to fail. Introducing a syntax error in any notebook code cell causes CI to fail. Both are demonstrated in the implementing PR.

### Phase 2 — Executable evidence

Adds verification the repository does not have today. Phase 2 items **SHOULD** be ordered before Phase 3 because they produce evidence rather than prose.

- **R2.1** The repository **MUST** contain a reproducible sample dataset under `examples/sample-data/` as a ZIP archive containing `train.csv` and `val.csv` conforming to `TABULAR_REGRESSION_DATASET_SPEC.md`.
- **R2.2** `examples/` **MUST** contain a build script that regenerates the archive deterministically from a documented public source.
- **R2.3** `examples/sample-data/DATASET_CARD.md` **MUST** document schema, provenance, licence, and row and column counts.
- **R2.4** The sample `train.csv` and `val.csv` targets **MUST** be non-constant, so the archive is a valid input under the target-validation policy established by PR #10.
- **R2.5** The repository **MUST** contain an artifact-inference notebook that loads `artifact.json`, `training_context.csv`, and the pinned checkpoint in a fresh process, and runs batch inference **without refitting**.
  *Rationale: fresh-process reload is currently listed as an unverified gate in the production acceptance boundary of `DIMER_CONTRACT.md`. This is the first artifact that exercises it.*
- **R2.6** The artifact-inference notebook **MUST** be covered by the R1.3 AST validator.
- **R2.7** Sample archives **MUST NOT** be presented as coverage for the dataset safety guards. Those guards are already exercised by archives constructed in `tests/test_dimer_runtime.py`, and synthetic construction is preferable for adversarial cases because it avoids committing a zip bomb to version control.

**Acceptance.** A reviewer regenerates the archive from the build script and obtains a byte-identical result, then runs the artifact-inference notebook against a completed run's artifacts and reproduces its reported metrics.

### Phase 3 — Operations documentation

- **R3.1** `DEPLOYMENT.md` **MUST** exist and **MUST** specify component split, image bases, network egress boundaries at build and run time, resource profile, and weights delivery.
- **R3.2** `DEPLOYMENT.md` **MUST** reference `DIMER_CONTRACT.md` for any fact already stated there, and **MUST NOT** duplicate it. Each fact has exactly one authoritative location.
- **R3.3** `weights/README.md` **MUST** document provenance, the pinned revision, the SHA-256 of `tabdpt1_2.safetensors`, a verification snippet, and offline staging instructions.
  *Note: this is not a parity item. No baseline has this file; TabICL's READMEs link to it and those links are broken. Implementing it here fixes a gap the baselines share.*
- **R3.4** `tutorials/README.md` **MUST** describe each notebook, its required accelerator, and its expected runtime.

**Acceptance.** An operator provisions the pipeline using only `DEPLOYMENT.md` and `weights/README.md`, with no access to this issue.

### Phase 4 — Tier split (blocked on D1, D3)

- **R4.1** A dataset-validator repository **MUST** exist, containing `Dockerfile`, `validate.py`, `requirements.txt`, and `tests/`.
- **R4.2** The validator image **MUST** run on CPU only and **MUST NOT** depend on `torch`, `tabdpt`, or the checkpoint.
- **R4.3** The validator **MUST** emit a `result.json` envelope at `DIMER_RESULT_PATH`, defaulting to `/data/dataset-validations/result.json`, for both pass and fail outcomes, including on internal error.
- **R4.4** The validator **MUST NOT** vendor a copy of the dataset guard implementations. It **MUST** obtain them by installing `tabdpt-regressor-pipeline` at a pinned git tag (per D3).
- **R4.5** The validator **MUST** reject, with a distinct machine-readable code per case: duplicate splits, nested archives, path traversal, symlinked roots, compression-ratio violations, file-count violations, malformed runtime configuration, `target_column` present in `drop_columns`, non-numeric or non-finite targets, and constant targets in a supplied `train.csv` or `val.csv`.
- **R4.6** A finetuner repository **MUST** exist, containing `Dockerfile`, `train.py`, `dimer-pipeline.json`, `requirements.txt`, and `tests/`.
- **R4.7** `dimer-pipeline.json` **MUST** remain byte-identical to this repository's copy, or a single authoritative copy **MUST** be designated and the other removed.
- **R4.8** Both satellite repositories **MUST** run CI on the same Python matrix as this repository (3.10–3.14).
  *Rationale: both baselines test a single interpreter version. Adopting their layout **MUST NOT** import their narrower CI.*

**Acceptance.** Both images build through DIMER. The validator passes the Phase 2 sample archive and rejects each adversarial case in R4.5 with the documented code. The finetuner completes a GPU smoke run and emits artifacts satisfying the existing `artifact.json` contract.

### Phase 5 — Serving-context format (blocked on D2)

Proposed as a standalone change. **MUST NOT** be bundled into Phases 1–4.

- **R5.1** If parquet is adopted, the artifact format identifier **MUST** increment to `tabdpt-dimer-context-v3`.
- **R5.2** The change **MUST** update `DIMER_CONTRACT.md`, `README.md`, `dimer_runtime.py`, and the by-path SHA-256 entries in `result.json` and `artifact.json` in one commit.
- **R5.3** `pyarrow` **MUST** be added as an explicit runtime dependency with a pinned major version.
- **R5.4** The artifact-inference notebook from R2.5 **MUST** be updated in the same change.

## 6. Risks

**Documentation divergence.** Four new documents across two repositories must stay consistent with two contracts. R3.2 mitigates by forbidding duplication.

**Parity for its own sake.** Two of the earlier analysis's items (`weights/README.md`, the three named sample datasets) describe artifacts that do not exist in any baseline. Requirements here are justified on this repository's needs, not on matching a file list.

**CI regression via layout adoption.** Adopting the TabICL layout risks importing its single-version CI. R4.8 forbids this.

## 7. Out of scope

Real 254 MB checkpoint execution in CI, physical T4 or Ampere provisioning in CI, on-platform DIMER deployment, and full TabDPT/Torch/FAISS compatibility testing across Python 3.10–3.14 remain platform-owned release gates, consistent with the existing production acceptance boundary in `DIMER_CONTRACT.md`.

## 8. Evidence

All claims in §1 and §2 were verified by direct inspection of cloned repositories at the SHAs current on 2026-09-08. Baseline repositories were read, not modified. Items marked `[TODO:]` are unresolved and **MUST** be settled before the phase that depends on them begins.

**Comment by kurtvalcorza** [comment]
### Milestone Update: Phases 1–3 Complete

Phases 1, 2, and 3 of this RFC have been implemented, verified, reviewed, and merged to `main` via PR #11 (commit `cd52f13`):

- **Phase 1 — Acceptance Definition & Tutorial Verification:**
  - `DEPLOYMENT.md` enumerates release gates distinguishing repository-verifiable steps from platform-owned gates (R1.1, R1.2).
  - `scripts/validate_colab_tutorial.py` enforces AST syntax validation, explicit `use_flash=False`, and forbids absolute local paths across tutorials (R1.3–R1.5).
  - GitHub Actions CI executes `validate_colab_tutorial.py` on push and PR across Python 3.10–3.14 (R1.6).

- **Phase 2 — Executable Evidence & Artifact Parity:**
  - `examples/sample-data/diabetes.zip` committed with deterministic builder script and `DATASET_CARD.md` (R2.1–R2.4).
  - Fresh-process artifact inference tutorial `tutorials/tabdpt_regressor_artifact_inference_colab.ipynb` added without refitting (R2.5–R2.6).
  - Upstream PCA basis, imputer statistics, and scaler parameters are persisted and restored with device and dtype preservation, preventing device mismatch on wide tables (>128 features). Verified by automated regression tests in `tests/test_artifact_reload.py`.

- **Phase 3 — Operations Documentation:**
  - `DEPLOYMENT.md` specifies component splits, container base images, network boundaries, and resource bounds (R3.1, R3.2).
  - `weights/README.md` documents weight SHA-256 verification and offline caching instructions (R3.3).
  - `tutorials/README.md` details hardware requirements and expected runtime for all notebooks (R3.4).

### Remaining Phases
- **Phase 4 (Tier Split):** Pending decisions D1/D3 regarding satellite repository creation (`tabdpt-regressor-dataset-validator`, `tabdpt-regressor-finetuner`).
- **Phase 5 (Serving Context Format):** Standalone evaluation of `tabdpt-dimer-context-v3` parquet format (D2).

**Comment by kurtvalcorza** [comment]
### Milestone Update: Phase 5 Complete & Phase 4 Deferred

#### Phase 5 — Serving-Context Format v3 (Merged to `main` via PR #13, commit `248f96c`)
Phase 5 has been fully implemented, verified, reviewed, and merged to `main`:
- **R5.1 Format Identifier:** Upgraded artifact format identifier to `tabdpt-dimer-context-v3` in `artifact.json` and `result.json`.
- **R5.2 Parquet Context Storage & Digest Integrity:** Emits `training_context.parquet` with verified SHA-256 digest in `artifact.json`.
- **R5.3 Dual-Format & Reader Hardening:**
  - `load_artifact()` supports both `tabdpt-dimer-context-v3` Parquet and legacy `tabdpt-dimer-context-v2` CSV bundles without refitting.
  - Case-insensitive extension dispatch (`.parquet`, `.PARQUET`, `.pq`) and explicit `engine="pyarrow"`.
  - Informative `ImportError` when `pyarrow` is absent.
  - Preserves categorical null markers (`<NA>`) via pandas nullable string dtype (`.astype("string")`), preventing missing values from collapsing into unknown-category codes during artifact reload.
- **R5.4 Pinned Runtime Dependency:** Constrained `pyarrow>=25.0,<26` strictly to a single validated major version in `pyproject.toml`.
- **Notebooks & Documentation:** Updated `tutorials/tabdpt_regressor_artifact_inference_colab.ipynb`, `DIMER_CONTRACT.md`, `DEPLOYMENT.md`, and `README.md`.
- **Automated Verification:** 49 passing tests, 1 skipped in `pytest`; full CI matrix green across Python 3.10–3.14.

#### Status of Remaining Phase
- **Phase 4 (Tier Split):** Deferred for now while `ml-worker` infrastructure is active. Decisions D1 (satellite-only architecture) and D3 (pipeline package reuse) will be enacted when Phase 4 container split work resumes.
